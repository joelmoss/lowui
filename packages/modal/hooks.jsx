import { createContext, useContext } from "react"

import useModel from "./model"

export const ModalContext = createContext({ id: undefined })

/**
 * @param {string | number} idFromArg
 * @returns {string | number}
 */
export const useModalId = (idFromArg) => {
  const { id: idFromContext } = useContext(ModalContext)
  return idFromArg || idFromContext
}

/**
 * Toggle the modal open or close
 * @callback ToggleModal
 * @param {boolean} isOpen
 */

/**
 * @param {string | number} idFromArg
 * @returns {{ id: string | number, isModalOpen: boolean, toggleModal: ToggleModal }}
 */
export const useModal = (idFromArg) => {
  const id = useModalId(idFromArg)
  const model = useModel()

  if (!id) throw new Error("useModal() requires a unique instance ID")

  return {
    id,
    isModalOpen: model.instances[id],
    toggleModal: (isOpen) => model.toggle(id, isOpen),
  }
}
