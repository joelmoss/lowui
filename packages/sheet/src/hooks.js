import { createContext, useContext } from 'react'

import useModel from './model'

export const SheetContext = createContext({ id: undefined })

export const useSheetId = idFromArg => {
  const { id: idFromContext } = useContext(SheetContext)
  return idFromArg || idFromContext
}

export const useSheet = idFromArg => {
  const id = useSheetId(idFromArg)
  const model = useModel()

  if (!id) throw new Error('useSheet() requires a unique instance ID')

  return {
    id,
    isSheetOpen: model.instances[id],
    toggleSheet: isOpen => model.toggle(id, isOpen)
  }
}
