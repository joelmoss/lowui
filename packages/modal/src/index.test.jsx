import { render, screen, waitFor } from "@testing-library/react"

import { useModal } from "./hooks"
import Modal from "."

test("closed by default", () => {
  render(
    <Modal id={1}>
      <div>withinTheModal</div>
    </Modal>
  )

  expect(screen.queryByText("withinTheModal")).not.toBeInTheDocument()
})

test("opening", async () => {
  const App = () => {
    const { toggleModal } = useModal(1)

    return (
      <>
        <button type="button" onClick={() => toggleModal(true)}>
          Open Modal
        </button>
        <Modal id={1}>
          <div>withinTheModal</div>
        </Modal>
      </>
    )
  }

  const { user } = setup(<App />)

  await user.click(screen.getByText("Open Modal"))

  expect(await screen.findByText("withinTheModal")).toBeInTheDocument()
})

test("canClose prop", async () => {
  render(
    <Modal id="canCloseProp" isOpen canClose={false}>
      <div>withinTheModal</div>
    </Modal>
  )

  await waitFor(() => {
    expect(screen.queryAllByRole("button")).toStrictEqual([])
  })

  expect(document.body).toMatchSnapshot()
})

test("isOpen prop", async () => {
  render(
    <Modal id={1} isOpen>
      withinTheModal
    </Modal>
  )

  expect(await screen.findByText("withinTheModal")).toBeInTheDocument()
})

test("onExit prop", async () => {
  const onExit = jest.fn()

  const App = () => {
    const { toggleModal } = useModal(1)

    return (
      <>
        <button type="button" onClick={() => toggleModal(false)}>
          Close Modal
        </button>
        <Modal id={1} isOpen onExit={onExit}>
          <div>withinTheModal</div>
        </Modal>
      </>
    )
  }

  const { user } = setup(<App />)

  await user.click(screen.getByText("Close Modal"))

  expect(onExit).toHaveBeenCalled()
})

test("multiple modals", async () => {
  const ModalOne = () => {
    return (
      <Modal id="modal1" isOpen>
        Modal1
      </Modal>
    )
  }

  const ModalTwo = () => {
    return <Modal id="modal2">Modal2</Modal>
  }

  const App = () => {
    return (
      <>
        <ModalOne />
        <ModalTwo />
      </>
    )
  }

  render(<App />)

  expect(await screen.findByText("Modal1")).toBeInTheDocument()
  expect(screen.queryByText("Modal2")).not.toBeInTheDocument()
})
