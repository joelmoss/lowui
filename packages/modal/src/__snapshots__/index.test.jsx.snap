// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`canClose prop 1`] = `
<body
  class=""
>
  <div />
  <div
    class="base"
  >
    <div
      class="dialog"
      role="dialog"
      tabindex="-1"
    >
      <div>
        withinTheModal
      </div>
    </div>
  </div>
</body>
`;
